src/System.Net.Security/src/System/Net/SecureProtocols/NegoState.Windows.cs

```diff
@@ -13,869 +13,614 @@
-                throw new InvalidOperationException(SR.net_auth_client_server);
+                throw new InvalidCredentialException(SR.net_auth_bad_client_creds, e);
             }
 
-            _exception = null;
-            _remoteOk = false;
-            _framer = new StreamFramer(_innerStream);
-            _framer.WriteHeader.MessageId = FrameHeader.HandshakeId;
+            if (e.NativeErrorCode == NegoState.ERROR_TRUST_FAILURE)
+            {
+                throw new AuthenticationException(SR.net_auth_context_expectation_remote, e);
+            }
 
-            _expectedProtectionLevel = protectionLevel;
-            _expectedImpersonationLevel = isServer ? impersonationLevel : TokenImpersonationLevel.None;
-            _writeSequenceNumber = 0;
-            _readSequenceNumber = 0;
+            throw new AuthenticationException(SR.net_auth_alert, e);
+        }
 
-            Interop.SspiCli.ContextFlags flags = Interop.SspiCli.ContextFlags.Connection;
+        private static bool IsLogonDeniedException(Exception exception)
+        {
+            Win32Exception win32exception = exception as Win32Exception;
 
-            // A workaround for the client when talking to Win9x on the server side.
-            if (protectionLevel == ProtectionLevel.None && !isServer)
-            {
-                package = NegotiationInfoClass.NTLM;
-            }
+            return (win32exception != null) && (win32exception.NativeErrorCode == (int)Interop.SecurityStatus.LogonDenied);
+        }
 
-            else if (protectionLevel == ProtectionLevel.EncryptAndSign)
-            {
-                flags |= Interop.SspiCli.ContextFlags.Confidentiality;
-            }
-            else if (protectionLevel == ProtectionLevel.Sign)
-            {
-                // Assuming user expects NT4 SP4 and above.
-                flags |= Interop.SspiCli.ContextFlags.ReplayDetect | Interop.SspiCli.ContextFlags.SequenceDetect | Interop.SspiCli.ContextFlags.InitIntegrity;
-            }
+        internal static Exception CreateExceptionFromError(SecurityStatusPal statusCode)
+        {
+            return new Win32Exception((int)SslStreamPal.GetInteropFromSecurityStatusPal(statusCode));
+        }
+
+        internal static int Encrypt(
+            SafeDeleteContext securityContext,
+            byte[] buffer,
+            int offset,
+            int count,
+            object secSizes,
+            bool isConfidential,
+            bool isNtlm,
+            ref byte[] output,
+            uint sequenceNumber)
+        {
+            SecSizes sizes = secSizes as SecSizes;
 
-            if (isServer)
+            try
             {
-                if (_extendedProtectionPolicy.PolicyEnforcement == PolicyEnforcement.WhenSupported)
-                {
-                    flags |= Interop.SspiCli.ContextFlags.AllowMissingBindings;
-                }
+                int maxCount = checked(Int32.MaxValue - 4 - sizes.BlockSize - sizes.SecurityTrailer);
```

Comment:
This will emit a NullReferenceException if the secSizes value had the wrong type.  Making me continue to question why it became `object` in the signature.